aiida/orm/utils.py

```diff
@@ -50,56 +50,93 @@ def WorkflowFactory(module):
+    elif inputs_provided > 1:
+        raise InputValidationError("only one of parameters 'node_id', 'pk' and 'uuid' "
                          "has to be supplied")
-    loaded_node = None
+
+    class_ = parent_class or Node
+    if not issubclass(class_,  Node):
+        raise TypeError("{} is not a subclass of {}".format(class_, Node))
+    # The logic is as follows: If pk is specified I will look for the pk
+    # if UUID is specified for the uuid.
+    # node_id can either be string -> uuid or an integer -> pk
+    # Checking first if node_id specified
     if node_id is not None:
-        if isinstance(node_id, str) or isinstance(node_id, unicode):
-            loaded_node = Node.get_subclass_from_uuid(node_id)
+        if isinstance(node_id, (str, unicode)):
+            uuid = node_id
         elif isinstance(node_id, int):
-            loaded_node = Node.get_subclass_from_pk(node_id)
-        else:
-            raise ValueError("'node_id' has to be either string, unicode or "
-                             "integer, {} given".format(type(node_id)))
-    if loaded_node is None:
-        if pk is not None:
-            loaded_node = Node.get_subclass_from_pk(pk)
+            pk = node_id
         else:
-            loaded_node = Node.get_subclass_from_uuid(uuid)
-
-    if parent_class is not None:
-        if not issubclass(parent_class, Node):
-            raise ValueError("parent_class must be a subclass of Node")
-        if not isinstance(loaded_node, parent_class):
-            raise NotExistent('No node found as '
-                              'subclass of {}'.format(parent_class))
-
-    return loaded_node
+            raise TypeError("'node_id' has to be either string, unicode or "
+                                 "integer, {} given".format(type(node_id)))
+
+            #I am checking whether uuid, if supplied,  is a string
+    if uuid is not None:
+        if not isinstance(uuid,(str, unicode)):
+            raise TypeError("'uuid' has to be string or unicode")
+        # or whether the pk, if provided, is an integer
+    elif pk is not None:
+        if not isinstance(pk, int):
+            raise TypeError("'pk' has to be an integer")
+    else:
+        # I really shouldn't get here
+        assert True,  "Neither pk  nor uuid was provided"
+
+    qb = QueryBuilder()
+    qb.append(class_, tag='node', project='*')
+
+    if pk:
+        qb.add_filter('node',  {'id':pk})
+    elif uuid:
+        # Put back dashes in the right place
+        start_uuid = uuid.replace('-', '')
+        if query_with_dashes:
+            # Essential that this is ordered from largest to smallest!
+            for dash_pos in [20,16,12,8]:
+                if len(start_uuid) > dash_pos:
+                    start_uuid = "{}-{}".format(
+                        start_uuid[:dash_pos], start_uuid[dash_pos:]
+                        )
+
+        qb.add_filter('node',{'uuid': {'like': "{}%".format(start_uuid)}})
+    try:
```

Comment:
I think the indent level of this is not 'ideal' - it's ok to work both if a pk or a uuid is given, but if a integer pk is given, this will give the wrong error message (actually, it will raise a different exception, because start_uuid is not defined.